### ISSUE TITLE
[Feature] Dynamic JS tab caption feature for the Tabs widget

### ISSUE THREAD

**Issue by daemons2000** [appsmithorg/appsmith#7359]
Please add a dynamic JS tab caption feature to the Tabs widget. Now tab captions are static.

Use case: Adding/hiding an asterisk in the tab caption dependent on whether some of the inputs on the tab have a non-empty value.

**Comment by Nikhil-Nandagopal** [comment]
@daemons2000 thank you for the feature request!

**Comment by somangshu** [comment]
@daemons2000 @Nikhil-Nandagopal not sure by what is meant by tab captions here, do we need to introduce this an a feature?

**Comment by Nikhil-Nandagopal** [comment]
@somangshu this is a label for the tab so that tab display names can be dynamic

**Comment by dilippitchika** [comment]
@somangshu since this is applicable to all widgets, do you want to track this in widget enhancements only and not table widget v2?

**Comment by marks0351** [comment]
yea the present field is dynamic in disguise which can solve the issue.

@somangshu @Nikhil-Nandagopal we need a better architecture to handle dynamic container type widgets(widgets that can have widgets inside them and can be created realtime like the List widget). 
we might need to get some inspiration from List Widget V2 and create a tabs widget v2 to make the tabs, tab labels, their visibility and order dynamic.
cc: @riodeuno @SatishGandham

**Comment by somangshu** [comment]
@marks0351 do you suggest we take this up later as v2 implementation, And not think about this is v1?

@dilippitchika I am confused
- how is applicable to all widgets
- why should it be tracked with the table v2

**Comment by marks0351** [comment]
@somangshu yes for now since we have a way to bind to tab value, we can leave it alone and think about how to make the tabs dynamic altogether(label, order, visibility).

**Comment by dilippitchika** [comment]
I ended up removing this from table v2, because it's not same. The thought was similar, that you have a different column name and a reference.

**Comment by danciaclara** [comment]
@somangshu @dilippitchika  Is there a separate feature to support Autocomplete on the Tab labels. A user was asking for it on discord. https://discord.com/channels/725602949748752515/760761686549463060/980435323911827486

**Comment by somangshu** [comment]
We need to identify what would it take to make this field support autocomplete, thats the only requirement we can pick up right now. @ohansFavour @aswathkk can you guys help me out?

**Comment by aswathkk** [comment]
@somangshu @dilippitchika  This field doesn't look like it supports binding. If we allow binding & auto-completion, then the next question will be why don't we support syntax highlighting and multi-line editing for this field. Have we taken a look at any different UX to make the tab label dynamic?

To support autocompletion here, we need to make use of the CodeEditor component instead of the Input component I guess. @ohansFavour correct me if I'm wrong.

**Comment by dilippitchika** [comment]
Agreed with @aswathkk i don't think we should add auto-completion support here as the component wasn't intended to be used this way. This was also the motivation behind adding the new label field. 

Checking the stats we have only 10 instances of people binding data in tab names. If we are picking it up we will add the label field and do it.

**Comment by Nikhil-Nandagopal** [comment]
I think we also need to think about the fundamental solution here. The name of the tab is a unique identifier we need internally and that is why we never supported binding to begin with. If we want to have a way to control this, we should think of a separate field like Display Name which can take a binding and is used to override the tab name. I think this same problem and confusion exists in table columns as well and it causes challenges in binding. The variable id we need for the column / tab should differ from the display name

**Comment by dilippitchika** [comment]
## Stats
| Stat | Values |
| ------------- | ------------- |
| Reach     | 65 |
| Effort (months)  | 0.5 |

**Comment by pchambless** [comment]
The App I am developing is a drill down scenario.  On the first tab, you select an Ingredient Type from a list.  This will present a table with the list of Ingredients of that Type in the 2nd tab.  Once an Ingredient in this table is selected, it will bring up a list of Ingredient batches for that ingredient.  On the 2nd Tab I am wanting the tab to display the Ingredient Type's name.  On the 3rd Tab I would like it to display the Ingredient Name like below.
![image](https://user-images.githubusercontent.com/44928133/227608899-72a6f268-2868-48c5-843b-a3e476c8ad98.png)
That being said, I'd like the ability to refer to each tab without needing to know what it is called, like **Tab.Index** or something similar

**Comment by somangshu** [comment]
@pchambless thanks for the request, this is a bit different from the original request. 

You want to set active tab with the help of something like index. cc @dilippitchika this is something that should be easy to implement, Can you see from a product POV, what this could otherwise imply, its impact etc!

**Comment by Zeral-Zhang** [comment]
I want use appsmith to build a support i18n product, so any label should use js binding to read from table or othre place, like table widget column header, tab widget title and so on.

https://github.com/user-attachments/assets/7fe2c8c3-2b58-48b3-b3cb-44ab815723fc


I found other low code platform like lowcoder, it supported this feature.
